src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -829,15 +834,16 @@ void ShenandoahConcurrentGC::op_final_mark() {
         }
 
         heap->set_evacuation_in_progress(true);
-        // From here on, we need to update references.
-        heap->set_has_forwarded_objects(true);
 
         // Verify before arming for concurrent processing.
         // Otherwise, verification can trigger stack processing.
         if (ShenandoahVerify) {
           heap->verifier()->verify_during_evacuation();
         }
 
+        // From here on, we need to update references.
+        heap->set_has_forwarded_objects(true);
+
```

Comment:
At both places, this makes sense since the collection set isn't empty, apropos of my earlier comment.

What is the reason to not have `set_has_forwarded_objects` being true at verification time? I notice that you made the same change in the earlier part as well. Would it be worth a documentation comment?

Reply:
Previously, the verifier treated "evacuation in progress" and "heap has forwarded objects" as the same state. We now have a case where "evacuation"  (promote-in-place) may happen without forwarded objects.